### PR TITLE
Adapted non-force fetching with locking

### PR DESCRIPTION
**Pull request by taruti**

### CODE REVIEW

**Comment by taruti** [comment]
And conflicts between src-d/master and keybase/master it seems.

**Comment by strib** [comment]
Can you please rebase this on top of `keybase/go-git/master`?  They're out of sync because of my outstanding status PR src-d/go-git#570.  It's hard to review with all the extra commits.

File: storage/filesystem/internal/dotgit/dotgit.go
Comment (strib) [inline storage/filesystem/internal/dotgit/dotgit.go]: Shouldn't there be a corresponding `f.Unlock()` somewhere, maybe as a `defer`?

File: options.go
Comment (strib) [inline options.go:150]: I know this is from someone else's PR, but I don't think this is strictly needed because the `RefSpecs` above already have an `IsForce()` method that says whether that particular ref should be forced or not.  (For `PullOptions`, the default `RefSpec` is defined in the `RemoteConfig`.)  Could you please update it to use the `IsForce` method instead?

**Review by strib — APPROVED** [review]
Actually I was able to review the relevant two patches separately.  Looks good with some minor suggestions, thanks!  But please rebase before merging into master.  The go-git folks are going to request some tests as well, so it would be good to write some, but that needn't block the merge here.

File: storage/filesystem/internal/dotgit/dotgit.go
Comment (taruti) [reply storage/filesystem/internal/dotgit/dotgit.go]: Actually Close takes care of it, but should probably document that. (If we were unlocking the file it would probably(?) need a fsync before the unlock, but go-billy lacks fsync too).

File: options.go
Comment (taruti) [reply options.go:150]: Will do.

File: storage/filesystem/internal/dotgit/dotgit.go
Comment (strib) [reply storage/filesystem/internal/dotgit/dotgit.go]: @taruti: I don't understand how `Close()` takes care of it.  Is that a new requirement of the `billy.File.Close()` method?  Our current implementation doesn't do that, though I guess maybe for posfix flock and Windows, `Close()` handles it automatically?  We will need something that explicitly calls `Unlock()` on our own `libfs.File` implementation; if `libfs.File.Close()` is supposed to do that, we can make that happen.  Was just confused.  A comment here would definitely help, and maybe one over in `go-billy` as well if that's a new assumption.

Comment (taruti) [reply storage/filesystem/internal/dotgit/dotgit.go]: But flocks are released by closing the file? From e.g. Linux ```Furthermore, the lock is released either by an explicit LOCK_UN operation on any of these duplicate file descriptors, or when all such file descriptors have been closed.```. But yes more documentation would be good.

Comment (taruti) [reply storage/filesystem/internal/dotgit/dotgit.go]: Our implementation of file locking will run into trouble with other applications if we keep the locks after all file descriptors are closed.

File: options.go
Comment (taruti) [reply options.go:150]: Now when actually doing this, are you sure we want this? CheckoutOptions and PullOptions have the Force bool in addition to FetchOptions. Removing it from one but not the others would feel inconsistent.

File: storage/filesystem/internal/dotgit/dotgit.go
Comment (strib) [reply storage/filesystem/internal/dotgit/dotgit.go]: The implementation of our KBFS locks won't be flock-based, so we'll need some way to know to release them.  Doing it as part of the `Close()` call is totally fine, we should just document it.

(@songgao -- we should change our `libfs.File.Close()` method to call `Unlock()`).

File: options.go
Comment (strib) [reply options.go:150]: You can keep the `Force` bool if you want, but in that case please also add support for the `IsForce()` method.  The git remote protocol supports both force and non-force fetches in the same batch, so we'll need to be able to specify the force on individual refs.

**Comment by taruti** [comment]
Made a rebased clear PR (to not mess up diffs here): https://github.com/keybase/go-git/pull/2

**Comment by taruti** [comment]
Closing this and merging the other one.

**Pull request closed without merging**
